cocotb/share/lib/verilator/verilator.cpp

```diff
@@ -0,0 +1,88 @@
+// Copyright cocotb contributors
+// Licensed under the Revised BSD License, see LICENSE for details.
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "Vtop.h"
+#include "verilated.h"
+#include "verilated_vpi.h"
+
+#include <memory>
+
+#if VM_TRACE
+# include <verilated_vcd_c.h>
+#endif
+
+vluint64_t main_time = 0;       // Current simulation time
+
+double sc_time_stamp() {       // Called by $time in Verilog
+    return main_time;           // converts to double, to match
+                                // what SystemC does
+}
+
+extern "C" {
+void vlog_startup_routines_bootstrap(void);
+}
+
+int main(int argc, char** argv) {
+    Verilated::commandArgs(argc, argv);
+#ifdef VERILATOR_SIM_DEBUG
+    Verilated::debug(99);
+#endif
+    std::unique_ptr<Vtop> top(new Vtop(""));
+    Verilated::fatalOnVpiError(false); // otherwise it will fail on systemtf
+
+    vlog_startup_routines_bootstrap();
+    VerilatedVpi::callCbs(cbStartOfSimulation);
+
+#if VM_TRACE
+    Verilated::traceEverOn(true);
+
+    std::unique_ptr<VerilatedVcdC> tfp(new VerilatedVcdC);
+    top->trace(tfp, 99);
```

Comment:
Does this compile? Or does this need to be `tfp.get()`?

Reply:
Yes, you are right, will open a PR for that.

Reply:
In fact, you don't need `unique_ptr` at all here - you can place these objects on the stack, replace `->` with `.`, and add a `&` here.

Reply:
Whats the advantage over the old, simple solution?

Reply:
ah, now I see what you mean. I actually would have liked it to stay consistent with verilator boilerplate code, but will amend

Reply:
> Verilator boilerplate code

Mind linking to this?

Reply:
https://www.veripool.org/projects/verilator/wiki/Manual-verilator#CONNECTING-TO-C

I think its generally okay to allocate on stack, but I am not sure about the implications it may have on large designs.

Reply:
I'd recommend updating those docs to use `unique_ptr`, assuming you're a verilator contributor.

Reply:
> I'd recommend updating those docs to use `unique_ptr`, assuming you're a verilator contributor.

Verilator hasn't made it to C++11 yet:
https://www.veripool.org/boards/3/topics/2580-Verilator-Requiring-C-11-compiler
One day . . .

Reply:
The described scenario probably doesn't even have a recent Python to run cocotb ;)

I think it makes sense to revert #1240 to just properly deleting..

Reply:
@wallento, are you saying that your comment at https://github.com/cocotb/cocotb/pull/943#issuecomment-565586936 was not yet true?